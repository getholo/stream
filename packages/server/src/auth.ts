import { ParameterizedContext } from 'koa';
import { createHash } from 'crypto';
import nanoid from 'nanoid';

type Next = () => Promise<any>

export const md5 = (value: string) => createHash('md5').update(value).digest('hex');

interface Digest {
  username: string
  realm: string
  nonce: string
  response: string
  cnonce: string
  nc: string
  qop: string
  uri: string
}

interface DigestAuthParams {
  header: string
  method: string
  realm: string
  password: string
}

export function digestAuth(params: DigestAuthParams) {
  const creds = params.header.replace('Digest', '').split(',').reduce(
    (obj, cred) => {
      const [name, value] = cred.split('=');
      return {
        ...obj,
        [name.trim()]: value.replace(/"/g, '').trim(),
      };
    },
    {} as Digest,
  );

  const hash1 = md5(`${creds.username}:${params.realm}:${params.password}`);
  const hash2 = md5(`${params.method}:${creds.uri}`);

  const digesthash = creds.qop === 'auth'
    ? md5([hash1, creds.nonce, creds.nc, creds.cnonce, creds.qop, hash2].join(':'))
    : md5([hash1, creds.nonce, hash2].join(':'));

  if (digesthash === creds.response) {
    return creds.username;
  }

  return undefined;
}

interface BasicAuthParams {
  header: string
  password: string
}

export function basicAuth(params: BasicAuthParams) {
  const creds = Buffer.from(params.header.replace('Basic', '').trim(), 'base64').toString('ascii');
  const [username, pass] = creds.split(':');

  if (pass === params.password) {
    return username;
  }

  return undefined;
}

export type AuthContext = ParameterizedContext<{ user: string }>

export const auth = (password: string, realm: string) => async (ctx: AuthContext, next: Next) => {
  const { request, response, method } = ctx;
  const header = request.get('authorization');

  if (!header) {
    response.set('WWW-Authenticate', `Digest realm="${realm}", nonce="${nanoid(24)}", algorithm="MD5"`);
    response.status = 401;
    return;
  }

  const isDigest = header.startsWith('Digest');
  const isBasic = header.startsWith('Basic');

  let username: string;

  if (isDigest) {
    try {
      username = digestAuth({
        header,
        method,
        password,
        realm,
      });
    } catch {
      // Incorrect header given, resolve to 401
    }
  }

  if (isBasic) {
    try {
      username = basicAuth({
        header,
        password,
      });
    } catch {
      // Set Header down below
    }
  }

  if (!username) {
    response.set('WWW-Authenticate', `Digest realm="${realm}", nonce="${nanoid(24)}", algorithm="MD5"`);
    response.status = 401;
  }

  ctx.state.user = username;
  await next();
};
