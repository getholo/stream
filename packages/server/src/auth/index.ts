import nanoid from 'nanoid';
import { IncomingMessage, ServerResponse } from 'http';
import Users from './users';

import { digest } from './digest';
import { basic } from './basic';

type Callback = (username: string) => any

const realm = 'beta@stream.getholo.dev';
const algorithm = 'MD5';
const qop = 'auth';

function notAuthenticated(res: ServerResponse) {
  const nonce = nanoid(24);
  res.setHeader('WWW-Authenticate', `Digest realm="${realm}",nonce="${nonce}",algorithm="${algorithm}",qop="${qop}"`);
  res.statusCode = 401;
  res.end();
}

export function withAuth(req: IncomingMessage, res: ServerResponse, users: Users, cb: Callback) {
  const header = req.headers.authorization;
  if (!header) {
    notAuthenticated(res);
    return;
  }

  const isDigest = header.startsWith('Digest');

  const username = isDigest
    ? digest({ header, users, method: req.method })
    : basic({ header, users });

  if (!username) {
    notAuthenticated(res);
    return;
  }

  cb(username);
}
