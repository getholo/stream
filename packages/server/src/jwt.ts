import base64url from 'base64url';
import crypto from 'crypto';

function signRS256(data: any, key: string) {
  const signer = crypto.createSign('RSA-SHA256');
  signer.update(data);
  const signature = signer.sign(key, 'base64');
  return base64url.fromBase64(signature);
}

export function createJWT(email: string, key: string) {
  const now = Math.floor(Date.now() / 1000);

  const header = base64url.encode(JSON.stringify({
    alg: 'RS256',
    typ: 'JWT',
  }));

  const payload = base64url.encode(JSON.stringify({
    iss: email,
    aud: 'https://oauth2.googleapis.com/token',
    scope: 'https://www.googleapis.com/auth/drive',
    iat: now,
    exp: now + 3600,
  }));

  const signature = signRS256(`${header}.${payload}`, key);
  return `${header}.${payload}.${signature}`;
}
