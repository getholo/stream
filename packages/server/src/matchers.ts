import pathToRegExp from 'path-to-regexp';

export type Resolution = 'best' | '2160' | '1080'

export function matchRoot(path: string) {
  const exp = pathToRegExp('/');

  const match = exp.exec(path);
  if (match) {
    return true;
  }

  return undefined;
}

const matchResolutionExp = pathToRegExp('/:resolution(1080|2160|best)');

export function matchResolution(path: string) {
  const match = matchResolutionExp.exec(path);
  if (match) {
    return {
      resolution: match[1] as Resolution,
    };
  }

  return undefined;
}

const matchFilmsExp = pathToRegExp('/:resolution(1080|2160|best)/films');

export function matchFilms(path: string) {
  const match = matchFilmsExp.exec(path);
  if (match) {
    return {
      resolution: match[1] as Resolution,
    };
  }

  return undefined;
}

const filmStreamExp = pathToRegExp('/:resolution(1080|2160|best)/films/:film');

export function matchFilmFile(path: string) {
  const match = filmStreamExp.exec(path);
  if (match) {
    return {
      resolution: match[1] as Resolution,
      film: match[2].split('.')[0],
    };
  }

  return undefined;
}

const matchShowsExp = pathToRegExp('/:resolution(1080|2160|best)/shows');

export function matchShows(path: string) {
  const match = matchShowsExp.exec(path);
  if (match) {
    return {
      resolution: match[1] as Resolution,
    };
  }

  return undefined;
}

const showFolderExp = pathToRegExp('/:resolution(1080|2160|best)/shows/:show');

export function matchShowFolder(path: string) {
  const match = showFolderExp.exec(path);
  if (match) {
    return {
      resolution: match[1] as Resolution,
      show: match[2],
    };
  }

  return undefined;
}

export function getSeason(seasonString: string) {
  const seasonMatch = /^(season\s)?(\d+)$/i.exec(seasonString);
  if (!seasonMatch) {
    return undefined;
  }

  return {
    season: parseInt(seasonMatch[2], 10),
    prefix: seasonMatch[1] || '',
  };
}

const seasonFolderExp = pathToRegExp('/:resolution(1080|2160|best)/shows/:show/:season');

export function matchSeasonFolder(path: string) {
  const match = seasonFolderExp.exec(path);

  if (!match) {
    return undefined;
  }

  const seasonMatch = getSeason(match[3]);
  if (!seasonMatch) {
    return undefined;
  }

  return {
    resolution: match[1] as Resolution,
    show: match[2],
    season: seasonMatch.season,
    prefix: seasonMatch.prefix,
  };
}

export function getEpisodeDetails(file: string) {
  const match = /S(\d+)E(\d+)/ig.exec(file);

  if (match) {
    return {
      season: parseInt(match[1], 10),
      episode: parseInt(match[2], 10),
    };
  }

  return undefined;
}

const episodeFileExp = pathToRegExp('/:resolution(1080|2160|best)/shows/:show/:season/:episode');

export function isEpisodeFile(path: string) {
  const match = episodeFileExp.exec(path);

  if (!match) {
    return undefined;
  }

  const seasonMatch = getSeason(match[3]);
  if (!seasonMatch) {
    return undefined;
  }

  const episodeDetails = getEpisodeDetails(match[4]);
  if (!episodeDetails) {
    return undefined;
  }

  return {
    resolution: match[1] as Resolution,
    show: match[2],
    season: seasonMatch.season,
    episode: episodeDetails.episode,
  };
}
