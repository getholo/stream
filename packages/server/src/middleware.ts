import { RouterContext } from '@koa/router';
import kleur from 'kleur';

import { createXML, createFolder } from './webdav';

export const propfind = (paths: string[]) => (ctx: RouterContext) => {
  const { request, response } = ctx;

  if (request.method === 'PROPFIND') {
    const folders = createXML(paths.map((path) => createFolder({ path })));

    response.type = 'text/xml';
    response.body = folders;
  }
};

// Analytics and Cache

export type AnalyticsContext = RouterContext<{
  user: string
  film: string
  resolution: '2160' | '1080'
  cached: boolean
}>;

export const analytics = async (ctx: AnalyticsContext, next) => {
  const start = Date.now();
  await next();

  if (!ctx.state.film) {
    return;
  }

  const { film, user, resolution } = ctx.state;
  const { status } = ctx.response;
  const range = ctx.response.get('content-range');

  console.log(`${kleur.magenta('Stream:')} ${user} is streaming: ${kleur.bold(film)} in ${resolution}p`);
  console.log(`> ${kleur.yellow(status)} ${kleur.dim(`${range} ${Date.now() - start}ms`)}`);
  console.log(`> ${kleur.dim(ctx.request.get('user-agent').split(' ')[0])}`);
};
