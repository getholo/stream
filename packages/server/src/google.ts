import axios from 'axios';

import { createJWT } from './jwt';

interface GoogleResponse {
  access_token: string
  expires_in: number
  token_type: string
}

interface Token {
  key: string
  expires: number
  token: string
}

let accessToken: Token;

export async function getAccessToken(email: string, key: string, force = false) {
  const now = Math.floor(Date.now() / 1000);

  if (!force && accessToken && now < accessToken.expires && accessToken.key === key) {
    return accessToken.token;
  }

  const { data } = await axios.request<GoogleResponse>({
    method: 'POST',
    url: 'https://oauth2.googleapis.com/token',
    data: {
      grant_type: 'urn:ietf:params:oauth:grant-type:jwt-bearer',
      assertion: createJWT(email, key),
    },
  });

  accessToken = {
    key,
    expires: now + 3200,
    token: data.access_token,
  };

  return accessToken.token;
}

export async function getStream(id: string, start: number, end: number, token: string) {
  const { data } = await axios.request<NodeJS.ReadableStream>({
    url: `https://www.googleapis.com/drive/v3/files/${id}?alt=media`,
    responseType: 'stream',
    method: 'GET',
    headers: {
      Authorization: `Bearer ${token}`,
      Range: `bytes=${start}-${end}`,
    },
  });

  return data;
}
