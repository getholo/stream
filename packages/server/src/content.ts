import axios from 'axios';
import { getAccessToken } from './google';
import { Files } from './data';
import { generateFilmExp, generateShowExp } from './paths';

export const extensions = {
  'video/quicktime': '.mov',
  'video/x-matroska': '.mkv',
  'video/mp4': '.mp4',
};

export interface DriveFile {
  id: string
  name: string
  mimeType: keyof typeof extensions | 'application/vnd.google-apps.folder'
  md5Checksum?: string
  parents: string[]
  size?: string
}

interface DriveFiles {
  files: DriveFile[]
  nextPageToken?: string
}

export async function fetchFiles(driveId: string, email: string, key: string) {
  const files: DriveFile[] = [];
  let pageToken: string = null;

  while (pageToken !== undefined) {
    const token = await getAccessToken(email, key);
    const { data } = await axios.request<DriveFiles>({
      method: 'GET',
      url: 'https://www.googleapis.com/drive/v3/files',
      params: {
        driveId,
        corpora: 'drive',
        pageSize: 1000,
        includeItemsFromAllDrives: true,
        supportsAllDrives: true,
        fields: 'files/md5Checksum,files/id,files/mimeType,files/parents,files/name,files/size,nextPageToken',
        q: "trashed = false and (mimeType = 'application/vnd.google-apps.folder' or mimeType = 'video/x-matroska' or mimeType = 'video/mp4')",
        pageToken,
      },
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });

    files.push(...data.files);
    pageToken = data.nextPageToken;
  }

  return files;
}

export interface FilmFile {
  mimeType: keyof typeof extensions
  id: string
  size: number
}

interface Versions {
  '2160'?: FilmFile
  '1080'?: FilmFile
}

export interface Films {
  [name: string]: Versions
}

export interface DriveParams {
  driveId: string
  email: string
  key: string
  files?: Files
  showRegex?: string
  filmRegex?: string
}

export class Content {
  private files: Files;

  private config: DriveParams;

  constructor(params: DriveParams) {
    this.config = {
      ...params,
      filmRegex: params.filmRegex || '/films/:film/:file',
      showRegex: params.showRegex || '/shows/:show/:season/:episode',
    };

    if (params.files) {
      this.files = params.files;
    } else {
      this.files = new Files();
    }
  }

  async firstFetch() {
    this.files.clear();

    const { driveId, email, key } = this.config;

    this.files.set(driveId, {
      name: 'Shared Drive',
      parent: null,
    });

    const allFiles = await fetchFiles(driveId, email, key);
    for (const file of allFiles) {
      this.files.set(file.id, {
        name: file.name,
        parent: file.parents[0],
        mimeType: file.mimeType,
        size: parseInt(file.size, 10),
      });
    }
  }

  createData() {
    const paths = this.files.paths();

    const { filmRegex, showRegex } = this.config;

    const films: Films = {};

    const matchFilm = generateFilmExp(filmRegex);
    const matchShow = generateShowExp(showRegex);

    for (const [id, path] of paths) {
      const isFilm = matchFilm(path);
      const isShow = matchShow(path); // <= future work

      if (isFilm) {
        const { file, film } = isFilm;
        const name = film.toLowerCase().replace(/ /g, '-').replace(/[^-\w]/g, '');
        const matchResolution = file.match(/((1080)|(2160))p/);
        const { mimeType, size } = this.files.get(id);

        if (matchResolution && (mimeType === 'video/mp4' || mimeType === 'video/x-matroska')) {
          const resolution = matchResolution[1] as '2160' | '1080';
          if (!films[name]) {
            films[name] = {};
          }

          films[name][resolution] = {
            id,
            mimeType,
            size,
          };
        }
      }
    }

    return {
      films,
    };
  }
}
