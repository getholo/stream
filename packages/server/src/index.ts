import Koa from 'koa';
import { Content, DriveParams } from './content';
import { Routing } from './routes';
import { auth } from './auth';

export { generateFilmExp, generateShowExp } from './paths';

interface Params extends DriveParams {
  password: string
  realm: string
}

export async function createServer(params: Params) {
  const content = new Content(params);
  await content.firstFetch();

  const app = new Koa();

  app.use(auth(params.password, params.realm));

  app.on('error', (error) => {
    if (error.code === 'EPIPE' || error.code === 'ECONNRESET') {
      return;
    }

    console.log(error);
  });

  const { films } = content.createData();
  const routing = new Routing({
    email: params.email,
    key: params.key,
    films,
  });
  app.use((ctx) => routing.middleware(ctx));

  return app;
}
