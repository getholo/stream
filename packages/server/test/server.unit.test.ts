import request from 'supertest';
import axios from 'axios';
import { IncomingMessage, ServerResponse } from 'http';

import { createServer } from '../src/index';

import { getAccessToken } from '../src/google';
import { DriveFile } from '../src/content';

jest.mock('../src/google');
jest.mock('axios');

const mockedAxios = axios as jest.Mocked<typeof axios>;
const getToken = getAccessToken as jest.Mock;

let callback: (req: IncomingMessage, res: ServerResponse) => void;

const files: DriveFile[] = [
  {
    id: '1',
    name: 'films',
    mimeType: 'application/vnd.google-apps.folder',
    parents: ['VA'],
  },
  {
    id: '2',
    name: 'joker-2019',
    mimeType: 'application/vnd.google-apps.folder',
    parents: ['1'],
  },
  {
    id: '3',
    name: 'joker.2019.2160p.mkv',
    mimeType: 'video/x-matroska',
    parents: ['2'],
    size: `${500 * 1024 * 1024}`,
  },
];

beforeAll(async () => {
  // mock mock, who's there?
  getToken.mockReturnValueOnce('token');
  // files
  mockedAxios.request.mockResolvedValueOnce({
    data: {
      files,
    },
  });

  // changeToken
  mockedAxios.request.mockResolvedValueOnce({
    data: {
      startPageToken: 'changeToken',
    },
  });

  const app = await createServer({
    driveId: 'VA',
    email: 'email',
    key: 'key',
    password: 'password',
    realm: 'test@getholo.dev',
  });

  callback = app.callback();
});

describe('Server Logic', () => {
  it('Path "/" should return 401 when not given Authorization header', async () => {
    const response = await request(callback)
      .propfind('/');

    expect(response.status).toEqual(401);
  });

  it('Path "/best/films" should return Joker when given Authorization header', async () => {
    const response = await request(callback)
      .propfind('/best/films')
      .auth('Test', 'password', { type: 'basic' });

    expect(response.status).toEqual(200);
    expect(response.type).toEqual('text/xml');
    expect(response.text).toMatchSnapshot();
  });
});
