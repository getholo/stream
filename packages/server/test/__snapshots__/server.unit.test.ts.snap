// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server Logic Path "/best/films" should return Joker when given Authorization header 1`] = `"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?><d:multistatus xmlns:d=\\"DAV:\\"><d:response><d:href>/best/films/joker-2019.mkv</d:href><d:propstat><d:prop><d:displayname>joker-2019.mkv</d:displayname><d:getcontentlength>524288000</d:getcontentlength><d:getcontenttype>video/x-matroska</d:getcontenttype></d:prop><d:status>HTTP/1.1 200 OK</d:status></d:propstat></d:response></d:multistatus>"`;
